File: android/src/main/kotlin/ru/kardenvan/flutter_yandex_ads_plus/platform_api/ad_event_dispatcher_facade/BasicAdEventDispatcherFacade.kt
package ru.kardenvan.flutter_yandex_ads_plus.platform_api.ad_event_dispatcher_facade

import com.yandex.mobile.ads.common.AdRequestError
import com.yandex.mobile.ads.common.ImpressionData
import ru.kardenvan.flutter_yandex_ads_plus.platform_api.ad_event.AdEvent
import ru.kardenvan.flutter_yandex_ads_plus.platform_api.ad_event_dispatcher.FlutterYandexAdsEventDispatcher

open class BasicAdEventDispatcherFacade(
    uid: String,
    dispatcher: FlutterYandexAdsEventDispatcher,
): AdEventDispatcherFacade(uid, dispatcher) {
    /**
     * Dispatches event notifying that ad was successfully loaded.
     */
    fun sendAdLoadedEvent() {
        sendBasicAdEvent(type = Type.ON_AD_LOADED)
    }

    /**
     * Dispatches event notifying that ad failed to load because of [error].
     */
    fun sendAdFailedToLoadEvent(error: AdRequestError) {
        sendBasicAdEvent(
            type = Type.ON_AD_FAILED_TO_LOAD,
            parameters = mapOf(
                "code" to error.code,
                "description" to error.description,
            )
        )
    }

    /**
     * Dispatches event notifying that ad called an impression that contains
     * provided [data].
     */
    fun sendImpressionEvent(data: ImpressionData?) {
        sendBasicAdEvent(
            type = Type.ON_IMPRESSION,
            parameters = mapOf(
                "data" to data?.rawData
            )
        )
    }

    /**
     * Dispatches event notifying that ad sent user to another application.
     */
    fun sendLeftApplicationEvent() {
        sendBasicAdEvent(type = Type.ON_LEFT_APPLICATION)
    }


    /**
     * Dispatches event notifying that user returned to current application after ad
     * sent him to another app.
     */
    fun sendReturnedToApplicationEvent() {
        sendBasicAdEvent(type = Type.ON_RETURNED_TO_APPLICATION)
    }

    /**
     * Dispatches event notifying that ad was clicked.
     */
    fun sendAdClickedEvent() {
        sendBasicAdEvent(type = Type.ON_AD_CLICKED)
    }

    /**
     * Dispatches event with given [uid], [type] and [parameters] to the platform.
     */
    private fun sendBasicAdEvent(
        type: Type,
        parameters: Map<String, Any?>? = null,
    ) {
        super.sendEvent(type.value, parameters)
    }

    private enum class Type constructor(
        val value: String
    ) {
        ON_AD_LOADED("onAdLoaded"),
        ON_AD_FAILED_TO_LOAD("onAdFailedToLoad"),
        ON_IMPRESSION("onImpression"),
        ON_LEFT_APPLICATION("onLeftApplication"),
        ON_RETURNED_TO_APPLICATION("onReturnedToApplication"),
        ON_AD_CLICKED("onAdClicked");
    }
}

File: ios/Classes/platform_api/ad_event_dispatcher_facade/BasicAdEventDispatcherFacade.swift
//
//  BasicAdEventDispatcher.swift
//  device_info_plus
//
//  Created by apple on 30.12.2022.
//

import Foundation

class BasicAdEventDispatcherFacade: AdEventDispatcherFacade {
    func sendOnLoadedEvent() {
        sendBasicAdEvent(
            type: BasicAdEventType.onAdLoaded
        )
    }

    func sendOnFailedToLoadEvent(error: Error) {
        sendBasicAdEvent(
            type: BasicAdEventType.onAdFailedToLoad,
            parameters: [
                "code": 0,
                "description": error.localizedDescription
            ]
        )
    }

    func sendOnClickedEvent() {
        sendBasicAdEvent(
            type: BasicAdEventType.onAdClicked
        )
    }

    func sendOnLeftApplicationEvent() {
        sendBasicAdEvent(
            type: BasicAdEventType.onLeftApplication
        )
    }
    
    func sendOnDidDismissScreenEvent() {
        sendBasicAdEvent(
            type: BasicAdEventType.didDismissScreen
        )
    }
    
    func sendOnWillPresentScreenEvent() {
        sendBasicAdEvent(
            type: BasicAdEventType.willPresentScreen
        )
    }
    
    func sendOnImpressionEvent(data: String?) {
        sendBasicAdEvent(
            type: BasicAdEventType.onImpression,
            parameters: [
                "data": data
            ]
        )
    }
    
    private func sendBasicAdEvent(
        type: BasicAdEventType,
        parameters: [String: Any?]? = nil
    ) {
        sendEvent(
            type: type.getValue(),
            parameters: parameters
        )
    }
    
    enum BasicAdEventType {
        case onAdLoaded
        case onAdFailedToLoad
        case onAdClicked
        case onLeftApplication
        case didDismissScreen
        case willPresentScreen
        case onImpression
        
        func getValue() -> String {
            switch self {
                case .onAdLoaded:
                    return "onAdLoaded"
                case .onAdFailedToLoad:
                    return "onAdFailedToLoad"
                case .onAdClicked:
                    return "onAdClicked"
                case .onLeftApplication:
                    return "onLeftApplication"
                case .didDismissScreen:
                    return "didDismissScreen"
                case .willPresentScreen:
                    return "willPresentScreen"
                case .onImpression:
                    return "onImpression"
            }
        }
    }
}
